### PR TITLE
Decorator do not have `data-offset-key` attribute. Fix `getSelectionOffsetKeyForNode` function

### PR DESCRIPTION
**Pull request by ouchxp**
Decorator do not have `data-offset-key` attribute. Fix the implementation of `getSelectionOffsetKeyForNode` function to resolve the `data-offset-key` from decorator's children and fix tests.

Related issue https://github.com/facebook/draft-js/issues/476

### CODE REVIEW

**Comment by hellendag** [comment]
Weird. What `SelectionState` is generated when double-clicking the link? What happens when stepping through https://github.com/facebook/draft-js/blob/master/src/component/selection/getDraftEditorSelectionWithNodes.js#L34?

**Comment by ouchxp** [comment]
@hellendag 
Using the example in this issue https://github.com/facebook/draft-js/issues/476

`global.getSelection()` returns 
```
Selection { anchorNode: <div.public-DraftStyleDefault-block.public-DraftStyleDefault-ltr>, anchorOffset: 1, focusNode: <div.public-DraftStyleDefault-block.public-DraftStyleDefault-ltr>, focusOffset: 2, isCollapsed: false, rangeCount: 1, caretBidiLevel: 0 }
```
Which indicates selection is the second child of the block, the link tag.

Inside `getDraftEditorSelectionWithNodes` function, it calls `getPointForNonTextNode` to get the `focusOffset`
Then in `getPointForNonTextNode` use `getSelectionOffsetKeyForNode` to find out the selected node is Leaf/Text or a Parent of Leaf. (This is the place it couldn't retrieve `data-offset-key` attribute from link decorator. So it consider the decorator as a Leaf/Text node.)

Finally, the `SelectionState` returned by `getDraftEditorSelectionWithNodes` is
```
Object { anchorKey: "9o6md", anchorOffset: 49, focusKey: "9o6md", focusOffset: 6, isBackward: true, hasFocus: true }
```

**Comment by ouchxp** [comment]
@hellendag I've check the Facebook Note, it does not have this problem because they added the `data-offset-key` to the link decorator. It's another way to make it work.

File: src/component/selection/__tests__/getDraftEditorSelection-test.js [hunk 1]
```diff
@@ -131,7 +131,8 @@ describe('getDraftEditorSelection', function() {
       .map(
         function(decoratorKey, ii) {
           var span = document.createElement('span');
-          span.setAttribute('data-offset-key', '' + decoratorKey);
+          // decorator do not have data-offset-key attribute
```
Comment (hellendag) [inline src/component/selection/__tests__/getDraftEditorSelection-test.js]:
```javascript
// Decorators may not have `data-offset-key` attribute
```

File: src/component/selection/getSelectionOffsetKeyForNode.js [hunk 2]
```diff
@@ -17,7 +17,15 @@
  * Get offset key from a node.
```
Comment (hellendag) [inline src/component/selection/getSelectionOffsetKeyForNode.js]: Probably a good idea to flesh out the docblock for this function, now that it does more than just check the current node.

File: src/component/selection/getSelectionOffsetKeyForNode.js [hunk 3]
```diff
@@ -17,7 +17,15 @@
...
  */
 function getSelectionOffsetKeyForNode(node: Node): ?string {
-  return node instanceof Element ? node.getAttribute('data-offset-key') : null;
+  if (node instanceof Element) {
+    var offsetKey = node.getAttribute('data-offset-key');
+    if (offsetKey) return offsetKey;
```
Comment (hellendag) [inline src/component/selection/getSelectionOffsetKeyForNode.js]: Styling nit. We should have eslint catch this.

```javascript
if (offsetKey) {
  return offsetKey;
}
```

Same for the `childOffsetKey` case below.

File: src/component/selection/getSelectionOffsetKeyForNode.js [hunk 4]
```diff
@@ -17,7 +17,15 @@
...
+    for (var ii = 0; ii < node.children.length; ii++) {
```
Comment (hellendag) [inline src/component/selection/getSelectionOffsetKeyForNode.js]: I'd prefer `childNodes` to `children` here, even though it means iterating over nodes that will return null when recursing.

**Comment by hellendag** [comment]
Thanks for the details. I think that sounds reasonable.

**Comment by ouchxp** [comment]
Thanks for the comments, I've fix these issues you've point out. Please feel free to leave more comments if there are still something need to be change.

**Comment by hellendag** [comment]
Cool, let's go with it.

**Pull request merged**